doc/diskd.1
.TH diskd 1 "03Mar05" fdutils-5.5
.SH Name
diskd - disk daemon; wait for disk to be inserted
'\" t
.de TQ
.br
.ns
.TP \\$1
..

.tr \(is'
.tr \(if`
.tr \(pd"

.SH Note
This manpage has been automatically generated from fdutils's texinfo
documentation.  However, this process is only approximative, and some
items, such as cross-references, footnotes and indices are lost in this
translation process.  Indeed, these items have no appropriate
representation in the manpage format.  Moreover, only the items specific
to each command have been translated, and the general information about
fdutils has been dropped in the manpage version.  Thus I strongly advise
you to use the original texinfo doc.
.TP
* \ \ 
To generate a printable copy from the texinfo doc, run the following
commands:
 
.nf
.ft 3
.in +0.3i
    ./configure; make dvi; dvips fdutils.dvi
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.TP
* \ \ 
To generate a HTML copy,  run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make html
.fi
.in -0.3i
.ft R
.lp
 
\&\fRA pre-made HTML can be found at:
\&\fR\&\f(CW\(ifhttp://www.tux.org/pub/knaff/fdutils\(is\fR
.TP
* \ \ 
To generate an info copy (browsable using emacs' info mode), run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make info
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.PP
The texinfo doc looks most pretty when printed or as HTML.  Indeed, in
the info version certain examples are difficult to read due to the
quoting conventions used in info.
.SH Description
.iX "p diskd"
.PP
The diskd command has the following syntax:
.PP
 
.nf
.ft 3
.in +0.3i
\&\fR\&\f(CWdiskd [\fR\&\f(CW-d \fIdrive\fR\&\f(CW] [\fR\&\f(CW-i \fIinterval\fR\&\f(CW] [\fR\&\f(CW-e \fIcommand\fR\&\f(CW]
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.PP
Diskd waits for a disk to be inserted into a given \fIdrive\fR, and then
either executes the \fIcommand\fR or exits. This program can be used to
automatically mount a disk as soon as it is inserted.
.PP
.SH Warning
.PP
This program works by switching the motor on for a very short
interval, and then seeking to track -1. This might damage hardware in
the long run. Amigas, which also use these techniques, are known for
having problems with their disk drives no longer spinning up properly
after a few month of usage.
.PP
.SH Options
.TP
\&\fR\&\f(CW-d\ \fIdrive\fR\&\f(CW\fR\ 
Selects the drive to observe for disk insertion. By default, drive 0
(\fR\&\f(CW/dev/fd0\fR) is observed.
.TP
\&\fR\&\f(CW-i\ \fIinterval\fR\&\f(CW\fR\ 
Selects the polling interval. The interval is given in tenths of
seconds. Default is 10 (one second).
.TP
\&\fR\&\f(CW-e\ \fIcommand\fR\&\f(CW\fR\ 
Gives the command to be executed when a disk is inserted. If no
command is given the program simply exits. Typically, the command
mounts the disk. It can be a shell scripts which probes for several
filesystems and disk geometries until it succeeds.
.PP
.SH Bugs
.IP
.TP
* \ \ 
Automatic unmounting cannot yet be handled. It is indeed not enough to
scan for disk removal, because when the disk is removed, it is already
too late: There might be some buffers needing flushing.  However, the
\&\fR\&\f(CWfdmountd\fR program allows automatic unmounting by using the
\&\fR\&\f(CWSYNC\fR mount options, which switches off write buffering
(see section  fdmount).
.TP
* \ \ 
The drive motor is running all the time, and on some computers, the
drive led flickers at each time the drive is polled.  
.SH See Also
Fdutils' texinfo doc


Changelog
020203	Fix fdmount and superformat bugs
020203	Create /dev/fd0 before testing for existence of /dev/fd0
011002	Fix floppycontrol.c printf conditional on FD_DISK_CHANGED, so that
	it works with gcc3 which apparently considers printf as a macro...
000612	Explicitly include sysmacros.h, apparently this is needed in
	certain environments
000603	Released 5.4
000527	Fixed a couple of buffer overrun exploits and other cleanup. Made
	fdutils compilable with Linux 2.3/2.4 kernels. Documentation fixes
970104	Fixed a few bugs listed in the Debian buglist posted to
	linux.debian.bugs.dis
	1.  Lessen dependancy of superformat on mtools by not failing if
	mformat cannot be called: low-level verify will still be done, but
	mbadblocks will be skipped
	2.  Fixed a few instances where the return value of getopt was
	misdeclared as a char instead of an integer
	3.  Fixed device names in fd.4
970411	Added error check to capacity measurement
	Corrected documentation Makefile bug
970629	Added vfat support to fdmount
970609	Fixed bug where a skew was erroneously applied to a 2m track
	Fixed typo in printfdprm
970604	Added vfat option for fdmount
970516	5.2 Released
970516	Fixed sunsite location in doc.
970430	Fixed typoes in doc about autodetection.
970428	Fixed manpage installation
970427	5.1 Released.  These repeated complaints about yylineno not
	defined start getting annoying ;-)
970426	More Makefile fixes.  Added note to doc about Alpha site
970422	Various Makefile fixes
970421	Lex yylineno fixes
970419	5.0 Released


src/diskd.c
#include <sys/types.h>
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <linux/fd.h>
#include <linux/fdreg.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <linux/major.h>
#include "enh_options.h"


int eioctl(int fd, int command,void * param, char *emsg)
{
  int r;
  if ((r=ioctl(fd,command,param))<0 )
    {
      perror(emsg);
      exit(1);
    }
  return r;
}

static char procFd[80];


int doPoll(int fd,
	   struct floppy_drive_params *dpr,
	   struct floppy_drive_struct *state)
{
	if(! (dpr->flags & FD_SILENT_DCL_CLEAR)) {
		/* Work around a bug in floppy driver when silent dcl is not
		   set */
		struct floppy_raw_cmd raw_cmd;
		int fd2;
		fd2=open(procFd, 3 | O_NDELAY);
		if(fd2 != -1)
			close(fd2);
		/* Perform "dummy" rawcmd to flush out newchange */
		raw_cmd.flags = FD_RAW_NEED_DISK;
		raw_cmd.cmd_count = 0;
		raw_cmd.track = 0;
		ioctl(fd, FDRAWCMD, &raw_cmd, "rawcmd");
	}
	eioctl(fd, FDPOLLDRVSTAT, state, "reset");
	return state->flags;
}

int main(int argc, char **argv)
{
	int mask=0;
	int fd=-2;
	char *command=0;
	int interval=10;
	int ch;
	struct timeval timval;
	struct floppy_drive_struct state;
	struct floppy_drive_params dpr;

	struct enh_options optable[] = {
	{ 'd', "drive", 1, EO_TYPE_FILE_OR_FD, 3, 0,
		  (void *) &fd,
		  "drive to be polled"},
		
	{ 'e', "exec", 1, EO_TYPE_STRING, 0, 0,
		  (void *) &command,
		  "shell command to be executed after disk insertion" },
		
	{ 'i', "interval", 1, EO_TYPE_LONG, 0, 0, 
		  (void *) &interval,
		  "set polling interval (in tenth of seconds)" },
	{ '\0', 0 }		
	};

	while((ch=getopt_enh(argc, argv, optable, 
			     0, &mask, "") ) != EOF ){
		if ( ch== '?' ){
			fprintf(stderr,"exiting\n");
			exit(1);
		}
		printf("unhandled option %d\n", ch);
		exit(1);
	}

	if ( fd == -2 )
		fd = open("/dev/fd0", 3 | O_NDELAY);
	if ( fd < 0 ){
		perror("can't open floppy drive");
		exit(1);
	}

	eioctl(fd,FDGETDRVPRM,(void *) &dpr, "Get drive parameters");
	sprintf(procFd, "/proc/self/fd/%d", fd);

	doPoll(fd, &dpr, &state);
	while (state.flags & (FD_VERIFY|FD_DISK_NEWCHANGE)) {
		timval.tv_sec = interval / 10;
		timval.tv_usec = (interval % 10) * 100000;
		select(0, 0, 0, 0, &timval);
		doPoll(fd, &dpr, &state);
	} 
	close(fd);
	if ( command)
		system(command);
	exit(0);
}
